// @antv/g6/lib/behavior/drag-canvas.js
var Util = require('../util');

var abs = Math.abs;
var DRAG_OFFSET = 10;
module.exports = {
  getDefaultCfg: function getDefaultCfg() {
    return {
      direction: 'both'
    };
  },
  getEvents: function getEvents() {
    return {
      'canvas:mousedown': 'onMouseDown',
      'canvas:mousemove': 'onMouseMove',
      'canvas:mouseup': 'onMouseUp',
      'canvas:click': 'onClick'
    };
  },
  updateViewport: function updateViewport(e) {
    var origin = this.origin;
    var clientX = +e.clientX;
    var clientY = +e.clientY;

    if (isNaN(clientX) || isNaN(clientY)) {
      return;
    }

    var dx = clientX - origin.x;
    var dy = clientY - origin.y;

    if (this.get('direction') === 'x') {
      dy = 0;
    } else if (this.get('direction') === 'y') {
      dx = 0;
    }

    this.origin = {
      x: clientX,
      y: clientY
    };
    this.graph.translate(dx, dy);
    this.graph.paint();
  },
  onMouseDown: function onMouseDown(e) {
    this.origin = {
      x: e.clientX,
      y: e.clientY
    };
    this.dragging = false;
  },
  onMouseMove: function onMouseMove(e) {
    e = Util.cloneEvent(e);
    var graph = this.graph;

    if (!this.origin) {
      return;
    }

    if (this.origin && !this.dragging) {
      if (abs(this.origin.x - e.clientX) + abs(this.origin.y - e.clientY) < DRAG_OFFSET) {
        return;
      }

      if (this.shouldBegin.call(this, e)) {
        e.type = 'dragstart';
        graph.emit('canvas:dragstart', e);
        this.dragging = true;
      }
    }

    if (this.dragging) {
      e.type = 'drag';
      graph.emit('canvas:drag', e);
    }

    if (this.shouldUpdate.call(this, e)) {
      this.updateViewport(e);
    }
  },
  onMouseUp: function onMouseUp(e) {
    if (!this.dragging) {
      this.origin = null;
      return;
    }

    e = Util.cloneEvent(e);
    var graph = this.graph;

    if (this.shouldEnd.call(this, e)) {
      this.updateViewport(e);
    }

    e.type = 'dragend';
    graph.emit('canvas:dragend', e);
    this.origin = null;
    this.dragging = false;
  },
  onClick: function onClick() {
    this.origin = null;
    this.dragging = false;
  }
};
